Sending rotation and height to external device over UART

rmeertens opened the pull request:
As I and other people in the MAVLab are working with computer vision on devices attached to the drone it would be nice to send the rotation and height of the drone to these devices. I added a module that does this.

kirkscheper commented:
There is some double code here with what is in modules/sensors/stereocam.c and .h. I would be more useful to merge those files with the stereoprotocol files you made so we only have one protocol to maintain.

Also I am not sure if subsystems is the best place for the stereoprotocol files, however I may not be the best person to say where it should go. Maybe subsystems/sensors or modules/sensors.

kirkscheper commented on sw/airborne/subsystems/stereoprotocol.h:
```diff
@@ -0,0 +1,60 @@
+ *      Author: roland
+ */
+
+#ifndef SW_AIRBORNE_SUBSYSTEMS_GPS_STEREOPROTOCOL_H_
```
This isn't the correct header format for this file location

kirkscheper commented on sw/airborne/subsystems/stereoprotocol.c:
```diff
@@ -0,0 +1,167 @@
+
+    for(uint8_t index=0; index < length; index++){
+        uint8_t toWrite = code[index];
+        dev->put_byte(dev->periph,code[index]);
```
You should check that there is space in the transmit buffer before trying to write to it with 

dev->check_free_space(struct uart_periph *p, uint8_t len);

Currently in the uart_arch ofr stm32 if the buffer is full it simply returns and doesn't give any indication that there is something wrong.

void uart_put_byte(struct uart_periph *p, uint8_t data)
{
  uint16_t temp = (p->tx_insert_idx + 1) % UART_TX_BUFFER_SIZE;

  if (temp == p->tx_extract_idx) {
    return;  // no room
  }
...

rmeertens commented:
Fixed almost all remarks by Kirk. 
I would prefer to put the protocol in some lib directory, but I couldn't find a suitable one. Anyone else has a preference?

Also: unrelated to the stereoprotocol I fixed some optical flow issues, and drones can now use the pix4flow to hover.

fvantienen commented:
Some remarks:
- Adding a ```RM``` directory isn't needed, since you don't use any personal airframes. (You are part of the TUDelft)
- Maybe take a look at periodic functions in modules, because the way you implemented the ```write_serial_rot``` function and the ```SERIALRMAT``` message shouldn't be the way to do this. Please remove the telemetry message ```SERIALRMAT``` and chang the ```write_serial_rot``` to a periodic function.
- For the stereoprotocol add the files to the folder of your module . Next to that you can also reuse original PPRZ message parsers and use that protocol(as we use almost everywhere in the lab).

Next to that I will add some remarks in the code itself.

fvantienen commented on conf/messages.xml:
```diff
@@ -2092,7 +2092,9 @@
 
   <!-- 245 is free -->
-  <!-- 246 is free -->
+ <message name="SERIALRMAT" id="246">
```
Remove this message, this is misusing the telemetry messages.

fvantienen commented on conf/modules/uartrotation.xml:
```diff
@@ -0,0 +1,22 @@
+    <file name="uartrotation.c"/>
+    <file name="stereoprotocol.c" dir="subsystems"/>
+	<raw>
+	ODROID_GPS_PORT_LOWER=$(shell echo $(GPS_PORT) | tr A-Z a-z)
```
GPS PORT?

fvantienen commented on conf/modules/uartrotation.xml:
```diff
@@ -0,0 +1,22 @@
+
+<module name="uartrotation">
+  <doc>
+    <description>Writes rotation to the uart bus</description>
```
Some more descriptive.. for Odroid only? which protocol?

fvantienen commented on conf/telemetry/default_rotorcraft.xml:
```diff
@@ -26,6 +26,7 @@
       <message name="AIR_DATA"               period="1.3"/>
       <message name="SURVEY"                 period="2.5"/>
       <message name="OPTIC_FLOW_EST"         period="0.25"/>
+      <message name="SERIALRMAT"         period="0.25"/>
```
Remove please..

fvantienen commented on sw/airborne/modules/optical_flow/px4flow.c:
```diff
@@ -40,11 +40,19 @@ bool_t optical_flow_available;
 static void decode_optical_flow_msg(struct mavlink_message *msg __attribute__((unused)))
 {
   optical_flow_available = TRUE;
+ // printf("Stabilisation decoded %d %d height: %f \n", optical_flow.flow_x, optical_flow.flow_y,optical_flow.ground_distance);
```
Remove commented out code

fvantienen commented on sw/airborne/modules/sensors/stereocam.c:
```diff
@@ -44,25 +44,16 @@ struct link_device *dev = STEREO_PORT;
 #define StereoSend(_dat,_len) { for (uint8_t i = 0; i< (_len); i++) StereoSend1(_dat[i]); };
```
This is old and shouldn't be nedded anymore

fvantienen commented on sw/airborne/modules/sensors/stereocam.c:
```diff
@@ -44,25 +44,16 @@ struct link_device *dev = STEREO_PORT;
 #define StereoSend(_dat,_len) { for (uint8_t i = 0; i< (_len); i++) StereoSend1(_dat[i]); };
 #define StereoUartSetBaudrate(_b) uart_periph_set_baudrate(STEREO_PORT, _b);
```
same here

fvantienen commented on sw/airborne/modules/uartrotation/uartrotation.c:
```diff
@@ -0,0 +1,65 @@
+#include "subsystems/datalink/telemetry.h"
+#include "subsystems/stereoprotocol.h"
+#include "navdata.h"
+int frameNumberSending=0;
```
Use static variables and preferable also structures.

fvantienen commented on sw/airborne/modules/uartrotation/uartrotation.h:
```diff
@@ -0,0 +1,32 @@
+ * reads from the odroid
+ */
+
+#ifndef READLOCATIONFROMODROID_H
```
Not really a correct name

fvantienen commented on sw/airborne/modules/uartrotation/uartrotation.h:
```diff
@@ -0,0 +1,32 @@
+ * <http://www.gnu.org/licenses/>.
+ */
+/**
+ * @file "modules/readlocationfromodroid/readlocationfromodroid.h"
```
Wrong path + file

fvantienen commented on sw/airborne/subsystems/stereoprotocol.c:
```diff
@@ -0,0 +1,164 @@
+/*
+ * stereoprotocol.cpp
```
Needs a correct header with copyright etc.

fvantienen commented on sw/airborne/subsystems/stereoprotocol.c:
```diff
@@ -0,0 +1,164 @@
+ * stereoprotocol.cpp
+ *
+ *  Created on: Sep 23, 2015
+ *      Author: roland
```
You're not the only one who made this...

fvantienen commented on sw/airborne/subsystems/stereoprotocol.c:
```diff
@@ -0,0 +1,164 @@
+//void stereoprot_get_msg_properties(uint8_t *, MsgProperties *, uint16_t,uint16_t);
+
+
+void WritePart(struct link_device *dev,uint8_t* code,uint8_t length){
```
Choose better function names please.. this is to generic.

fvantienen commented on sw/airborne/subsystems/stereoprotocol.h:
```diff
@@ -0,0 +1,60 @@
+/*
```
Same for header here

fvantienen commented on sw/airborne/subsystems/stereoprotocol.h:
```diff
@@ -0,0 +1,60 @@
+  uint8_t width;
+  uint8_t height;
+} ;
+typedef struct MsgProperties MsgProperties;
```
We don't prefer typedefs.. next to that if you do use a typedef use it before the struct.

fvantienen commented on sw/airborne/modules/uartrotation/uartrotation.c:
```diff
@@ -0,0 +1,65 @@
+#include <inttypes.h>
+#include <errno.h>
+#include "state.h"
+#include "subsystems/gps.h"
```
are all these includes needed????

fvantienen commented on sw/airborne/modules/uartrotation/uartrotation.c:
```diff
@@ -0,0 +1,65 @@
+/**
+ * @file "modules/readlocationfromodroid/readlocationfromodroid.c"
+ * @author Roland
+ * reads from the odroid and sends information of the drone to the odroid using JSON messages
```
JSON?

fvantienen commented on sw/airborne/modules/uartrotation/uartrotation.c:
```diff
@@ -0,0 +1,65 @@
+	pointer[10]=frameNumberSending++;
+	printf("Whoo sending serial\n");
+	stereoprot_sendArray( &((UART_LINK).device),ar, lengthArrayInformation, 1);
+	printf("Whoo sending serial2\n");
```
Remove debug code

fvantienen commented on sw/airborne/modules/uartrotation/uartrotation.c:
```diff
@@ -0,0 +1,65 @@
+
+static void write_serial_rot(struct transport_tx *trans, struct link_device *devasdf) {
+	struct Int32RMat *ltp_to_body_mat = stateGetNedToBodyRMat_i();
+	int32_t lengthArrayInformation = 11*sizeof(int32_t);
```
why 11? you can just to sizeof(struct Int32RMat) which is more clear. Also make it static.

fvantienen commented on sw/airborne/modules/uartrotation/uartrotation.c:
```diff
@@ -0,0 +1,65 @@
+	int32_t lengthArrayInformation = 11*sizeof(int32_t);
+	uint8_t ar[lengthArrayInformation];
+	int32_t *pointer = (int32_t*)ar;
+	for(int indexRot = 0; indexRot < 9; indexRot++){
```
Copying is not needed since the RMAT is already an array in the memory

fvantienen commented:
Please also squash your commits and use the codestyle formatter

gautierhattenberger commented on sw/airborne/modules/optical_flow/px4flow.c:
```diff
@@ -40,11 +40,18 @@ bool_t optical_flow_available;
 
 // request struct for mavlink decoder
 struct mavlink_msg_req req;
+#define PIX4FLOWSENDER_ID 1333
```
ABI sender id should be in subsystems/abi_sender_ids.h file and should be between 0 and 255

gautierhattenberger commented on sw/airborne/modules/optical_flow/px4flow.c:
```diff
@@ -40,11 +40,18 @@ bool_t optical_flow_available;
+  // Y negated to get to the body of the drone
+  AbiSendMsgVELOCITY_ESTIMATE(PIX4FLOWSENDER_ID, 0,
+                                  (optical_flow.flow_x/optical_flow.ground_distance),
+								  -1.0*(optical_flow.flow_y/optical_flow.ground_distance),
```
please check indentation

gautierhattenberger commented:
The files involved in the stereo cam sensor are scattered  in different directories with name that don't always make sense. Would it be possible to gather them at the same place ? Like a `stereo_cam` folder, even if some files could be reused in other situations.

In general, why the standard PPRZ format is not used to encode and send/receive messages ? For the stereo cam, I can agree that a specific sensor use a specific protocol, but when dealing with a generic external CPU like Odroid, why not using regular messages ? We now provide a Python interface and a C library can be extracted from the airborne code (this last stuff might be a bit broken, but not that much).

flixr commented:
I agree with @gautierhattenberger that unless there is a really good reason to implement a separate protocol the standard paparazzi protocol should be used to send messages to other onboard computers (like the odroid).
Also the protocol should not be limited to UART...

Also it would be nice to split unrelated changes into several pull requests (px4fow and sending attitude to odroid).
And hopefully you can find a better name than uartrotation...

rmeertens commented:
Fixed the ABI sender ID. 

The reason that the PPRZ protocol was not used was because I did not know it existed. As this protocol was used on our stereo boards for the communication for a long time I decided to add it to paparazzi. 
I still like to switch to the PPRZ protocol, but this will come later (including Felix's suggestion to add other ports than UART). 

Also: @gautierhattenberger , what would you like to have in this stereo_cam folder? The stereocam.h and .h files that are currently in the modules/sensors folder, the uart rotation stuff I added and the protocols?

knmcguire commented:
This would be very useful for me as well. Right now I'm implementing optical flow and divergences for the stereocam, and it would improve the quality if I could send IMU values to the stereoboard too.

gautierhattenberger commented:
@rmeertens Yes, the folder should have all files related to the stereo_cam sensor: main code and header, protocol and imu data sending. As @flixr suggested, a better name should be find for uavrotation. I would even suggect to merge it in the main code since their is only one function.
Concerning communication with external CPU, we have better solution using the Paparazzi message system. You can find an example with the `meteo_france_DAQ` module with a bidirectional communication with a Mbed MCU.

OpenUAS commented:
@rmeertens as Gautier proposed and a quick look at 
https://github.com/paparazzi/paparazzi/blob/master/sw/airborne/modules/meteo/meteo_france_DAQ.c

We know, implementing it in such a way sadly is work for you for now but likely will save work in the future. A renewed pull request after that is best. Will help to make it so. 

But first discuss it in gitter with core developers so it is clear that if implemented the next pull request can simply be honored quickly. Maybe something for Volker to fix.

rmeertens commented:
Moved the files to the stereo_cam folder, fixed the codestyle and tested my modules. In this pull request I am NOT going to add the PPRZ protocol. The code works right now and is useful for others. Although using the PPRZ format over serial is a great idea I would definitely have used it if it was better implemented (i.e. a git submodule that I could add to my other projects) and if I had known this before I implemented this. 
I would still like to use the PPRZ protocol for the communication between our stereo board and our drones, but I am not going to implement it in the coming month (also, @OpenUAS, implementing it is not really @Pold87 's task or expertise).

gautierhattenberger commented on sw/airborne/subsystems/abi_sender_ids.h:
```diff
@@ -226,4 +226,13 @@
 #endif
 
+#ifndef UART_ROTATION_ID 
+#define UART_ROTATION_ID 17
```
This ID is not used (and I don't see where it should be used)

gautierhattenberger commented on sw/airborne/modules/optical_flow/px4flow.c:
```diff
@@ -41,10 +41,17 @@ bool_t optical_flow_available;
+
+  // Y negated to get to the body of the drone
+  AbiSendMsgVELOCITY_ESTIMATE(PIX4FLOW_VELOCITY_ID, 0,
+                                  (optical_flow.flow_x/optical_flow.ground_distance),
```
indentation

gautierhattenberger commented:
We can probably merge this for know, but the uavrotation module is clearly not a long term solution. It shall be removed as soon as we can provide an easier to use API for messages.
